### PR TITLE
Move all mail related codes from models to services/mailer

### PR DESCRIPTION
**Pull request by lunny**
This refactor creates a new interface `DBContext` which could be used outside of `models`. You could use them on `cmds`, `routers` and `services`.

There are two connivence functions `WithContext` and `WithTransaction` that could be invoked.

This refactor also introduced a new top package named `services` and the first sub package `services/mailer` moved from `modules/mailer`.

### CODE REVIEW

File: modules/mailer/mail_issue.go [hunk 1]
```diff
@@ -118,39 +118,41 @@ func mailIssueCommentToParticipants(e Engine, issue *Issue, doer *User, content
...
 	mentions := markup.FindAllMentions(issue.Content)
 
-	if err = UpdateIssueMentions(e, issue.ID, mentions); err != nil {
+	if err = models.UpdateIssueMentions(e, issue.ID, mentions); err != nil {
 		return fmt.Errorf("UpdateIssueMentions [%d]: %v", issue.ID, err)
 	}
 
 	if len(issue.Content) > 0 {
-		if err = mailIssueCommentToParticipants(e, issue, doer, issue.Content, nil, mentions); err != nil {
+		if err = mailIssueCommentToParticipants(issue, doer, issue.Content, nil, mentions); err != nil {
```
Comment (zeripath) [inline modules/mailer/mail_issue.go]: You've dropped the engine from this function which basically means that passing in the engine in models.UpdateIssueMentions sort of pointless - you can't do this all in one transaction anyway. So at this point we could just drop the engine

Comment (lunny) [reply modules/mailer/mail_issue.go]: Yes, it should not be in a transaction. Sending mail failed should only warn user or log.

File: models/models.go [hunk 2]
```diff
@@ -139,6 +139,26 @@ func init() {
 	}
 }
 
+// WithEngine represents executing database operations
+func WithEngine(f func(e Engine) error) error {
```
Comment (zeripath) [inline models/models.go]: OK, although this is kind of what I was thinking we should do - it's not quite what I was thinking.

By passing in a `models.Engine` you're allowing arbitrary use of the xorm API outside of models - which I don't think we should permit. Use of xorm and the db probably should be restricted to functions in models.

I think we should be passing in a context (or at least a context provider). Something like:

```go
type DBContext struct {
  e Engine
}
```

then we could have `WithoutTransaction` and `WithTransaction` callback functions that set the `DBContext` appropriately. Then everywhere that uses `models.x` would be called within at least a `WithoutTransaction`. 

A better alternative would be through the use of a DBContextProvider interface:

```go
type DBContextProvider interface {
    GetDBContext() DBContext
}
```

We make our Gitea and API contexts a `DBContextProvider` by setting their `DBContext` to be say the `models.DefaultDBContext` - which simply returns `models.x` as the little `e` Engine. We then make any call to an exported function in `models` expect a `DBContextProvider` as the first argument.

`models.WithTransaction()` then provides a callback with a different `DBContextProvider` which has a transaction in it.

**Comment by lunny** [comment]
@zeripath done.

File: models/issue_user.go [hunk 3]
```diff
@@ -94,22 +94,22 @@ func UpdateIssueUserByRead(uid, issueID int64) error {
 }
 
 // UpdateIssueUsersByMentions updates issue-user pairs by mentioning.
-func UpdateIssueUsersByMentions(e Engine, issueID int64, uids []int64) error {
+func UpdateIssueUsersByMentions(ctx DBContext, issueID int64, uids []int64) error {
```
Comment (zeripath) [inline models/issue_user.go:97]: This should be in a transaction or should start its own

Comment (lunny) [reply models/issue_user.go:97]: It could be, but not from the function itself.

**Comment by lunny** [comment]
Ready to review.

File: services/mailer/mail.go [hunk 4]
```diff
@@ -98,14 +98,18 @@ func SendActivateEmailMail(locale Locale, u *User, email *EmailAddress) {
...
 	msg.Info = fmt.Sprintf("UID: %d, activate email", u.ID)
 
-	mailer.SendAsync(msg)
+	SendAsync(msg)
 }
 
 // SendRegisterNotifyMail triggers a notify e-mail by admin created a account.
-func SendRegisterNotifyMail(locale Locale, u *User) {
+func SendRegisterNotifyMail(locale Locale, u *models.User) {
+	if setting.MailService == nil {
```
Comment (techknowlogick) [inline services/mailer/mail.go:111]: Can you add Gitea to copyright?

**Review by techknowlogick — APPROVED** [review]
One tiny change required regarding copyright, but otherwise LGTM

**Comment by lunny** [comment]
@techknowlogick done.

**Comment by lunny** [comment]
@zeripath It's ready to review.

File: models/context.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// Insert inserts a object to database
+func (ctx *DBContext) Insert(obj interface{}) error {
+	_, err := ctx.e.Insert(obj)
+	return err
+}
+
+// LoadByID loads record from database according id, if it's not exist return ErrNotExist
+func (ctx *DBContext) LoadByID(id int64, obj interface{}) error {
```
Comment (guillep2k) [inline models/context.go]: I don't think this kind of interaction should be exposed outside models. I'd prefer to create a by-pass function when needed, like:
```
type DBContext struct {
	e Engine
}
func UpdateIssueMentionsCtx(ctx DBContext, issueID int64, mentions []string) error {
    return UpdateIssueMentions(ctx.Engine, issueID, mentions)
}
func UpdateIssueMentions(e Engine, issueID int64, mentions []string) error {
    [...]
}
```

Comment (lunny) [reply models/context.go]: I will follow that idea.

Comment (lunny) [reply models/context.go]: @guillep2k done.

File: models/error.go [hunk 6]
```diff
@@ -11,6 +11,21 @@ import (
...
+}
+
+// IsErrNotExist checks if an error is an ErrNotExist
+func IsErrNotExist(err error) bool {
+	_, ok := err.(ErrNotExist)
+	return ok
+}
+
+func (err ErrNotExist) Error() string {
+	return fmt.Sprintf("record is not exist [id: %d]", err.ID)
```
Comment (guillep2k) [inline models/error.go]:
```suggestion
	return fmt.Sprintf("record does not exist [id: %d]", err.ID)
```

File: services/mailer/mail.go [hunk 7]
```diff
@@ -99,14 +100,18 @@ func SendActivateEmailMail(locale Locale, u *User, email *EmailAddress) {
...
```
Comment (guillep2k) [inline services/mailer/mail.go:111]: Perhaps log an error if there's no mailer? This is an important mail.

**Comment by lunny** [comment]
@guillep2k all done.

**Review by guillep2k — APPROVED** [review]

**Pull request merged**